broker/aws/impl.go

```diff
@@ -160,10 +160,48 @@ func (b *Broker) getAWSRolesForAccount(accountName string) ([]string, error) {
 	return b.withSessionGetAWSRoleList(accountSession)
 }
 
+const roleCacheDuration = time.Second * 1800
+
+func (b *Broker) getAWSRolesForAccount(accountName string) ([]string, error) {
+	b.accountRoleMutex.Lock()
+	cachedEntry, ok := b.accountRoleCache[accountName]
+	b.accountRoleMutex.Unlock()
+	if ok {
+
+		if cachedEntry.Expiration.After(time.Now()) {
+			b.logger.Debugf(1, "GOT rolesfrom cache")
+			return cachedEntry.Roles, nil
+		}
+
+		// entry is expired
+		value, err := b.getAWSRolesForAccountNonCached(accountName)
+		if err != nil {
+			b.logger.Printf("Failure gettting non-cached roles, using expired cache")
+			return cachedEntry.Roles, nil
+		}
+		cachedEntry.Roles = value
+		cachedEntry.Expiration = time.Now().Add(roleCacheDuration)
+		b.accountRoleMutex.Lock()
+		b.accountRoleCache[accountName] = cachedEntry
+		b.accountRoleMutex.Unlock()
+		return value, nil
+	}
+	value, err := b.getAWSRolesForAccountNonCached(accountName)
+	if err != nil {
+		return value, err
+	}
+	cachedEntry.Roles = value
+	cachedEntry.Expiration = time.Now().Add(roleCacheDuration)
+	b.accountRoleMutex.Lock()
+	b.accountRoleCache[accountName] = cachedEntry
+	b.accountRoleMutex.Unlock()
+	return value, nil
+}
+
 func stringIntersectionNoDups(set1, set2 []string) (intersection []string) {
 	for _, v1 := range set1 {
 		for _, v2 := range set2 {
-			if v1 == v2 {
+			if strings.ToLower(v1) == strings.ToLower(v2) {
 				intersection = append(intersection, v1)
 			}
 		}
```

Comment:
This is an N^2 algorithm. How large will these slices be? Unless you're certain that these will always be small (in other organisations too), please convert one of these into a `map[string]struct{}` (i.e. a proper set) and then iterate over only one of the slices.